Applications/Spire/Source/Ui/DecimalInputWidget.cpp

```diff
@@ -0,0 +1,269 @@
+    static auto width = scale_width(8);
+    return width;
+  }
+
+  auto BUTTON_PADDING() {
+    static auto padding = scale_width(15);
+    return padding;
+  }
+}
+
+DecimalInputWidget::DecimalInputWidget(double value, QWidget* parent)
+    : QDoubleSpinBox(parent),
+      m_last_valid_value(value),
+      m_last_cursor_pos(0),
+      m_has_first_click(false) {
+  setValue(m_last_valid_value);
+  setDecimals(DEFAULT_DECIMAL_PLACES);
+  set_stylesheet(false, false);
+  setContextMenuPolicy(Qt::NoContextMenu);
+  connect(this, QOverload<double>::of(&QDoubleSpinBox::valueChanged), this,
+    &DecimalInputWidget::on_value_changed);
+  connect(lineEdit(), &QLineEdit::textEdited, this,
+    &DecimalInputWidget::on_text_edited);
+  lineEdit()->installEventFilter(this);
+}
+
+QString DecimalInputWidget::textFromValue(double value) const {
+  if(value == 0.0) {
+    return "0";
+  }
+  QString str = QString::number(value, 'f', decimals());
+  str.remove(QRegExp("0+$"));
+  str.remove(QRegExp("\\.$"));
+  return str;
+}
+
+connection DecimalInputWidget::connect_change_signal(
+    const ValueSignal::slot_type& slot) const {
+  return m_change_signal.connect(slot);
+}
+
+connection DecimalInputWidget::connect_submit_signal(
+    const ValueSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool DecimalInputWidget::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == lineEdit() && event->type() == QEvent::MouseButtonPress &&
+      !m_has_first_click && isEnabled()) {
+    auto e = static_cast<QMouseEvent*>(event);
+    if(e->button() == Qt::LeftButton) {
+      m_has_first_click = true;
+      selectAll();
+      return true;
+    }
+  }
+  return QDoubleSpinBox::eventFilter(watched, event);
+}
+
+void DecimalInputWidget::focusInEvent(QFocusEvent* event) {
+  if(event->reason() != Qt::MouseFocusReason) {
+    m_has_first_click = true;
+  }
+  QDoubleSpinBox::focusInEvent(event);
+}
+
+void DecimalInputWidget::focusOutEvent(QFocusEvent* event) {
+  m_has_first_click = false;
+  if(text().isEmpty()) {
+    blockSignals(true);
+    setValue(m_last_valid_value);
+    blockSignals(false);
+  }
+  QDoubleSpinBox::focusOutEvent(event);
+  m_last_valid_value = value();
+  m_submit_signal(m_last_valid_value);
+}
+
+void DecimalInputWidget::keyPressEvent(QKeyEvent* event) {
+  switch(event->key()) {
+    case Qt::Key_Delete:
+      blockSignals(true);
+      setValue(m_last_valid_value);
+      blockSignals(false);
+      return;
+    case Qt::Key_Enter:
+    case Qt::Key_Return:
+      if(text().isEmpty()) {
+        blockSignals(true);
+        setValue(m_last_valid_value);
+        blockSignals(false);
+      }
+      lineEdit()->setText(textFromValue(value()));
+      m_last_valid_value = value();
+      m_submit_signal(m_last_valid_value);
+      return;
+    case Qt::Key_Up:
+      add_step(1, event->modifiers());
+      return;
+    case Qt::Key_Down:
+      add_step(-1, event->modifiers());
+      return;
```

Comment:
Just to confirm, should all these branches bypass the call `QDoubleSpinBox::keyPressEvent(event)` below?

Reply:
Yes, the base class keyPressEvent() will have some kind of side effect (like highlighting all of the text) that I'm trying to avoid.